core/test/TreasureGrabTests.java

```diff
@@ -0,0 +1,192 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class TreasureGrabTests {
+
+    int treasureCounter = 0;
+    private Logic.MoveDirection[] parseMoveSequence(final String code) {
+        final char[] chars = code.toCharArray();
+        final Logic.MoveDirection[] dirs = new Logic.MoveDirection[code.length()];
+
+        for (int i = 0; i < dirs.length; i++) {
+            switch (chars[i]) {
+                case 'U':
+                    dirs[i] = Logic.MoveDirection.UP;
+                    break;
+                case 'L':
+                    dirs[i] = Logic.MoveDirection.LEFT;
+                    break;
+                case 'D':
+                    dirs[i] = Logic.MoveDirection.DOWN;
+                    break;
+                case 'R':
+                    dirs[i] = Logic.MoveDirection.RIGHT;
+                    break;
+                default:
+                    throw new IllegalArgumentException(
+                            String.format("Illegal char '%c' at %d", chars[i], i)
+                    );
+            }
+        }
+
+        return dirs;
+    }
+
+    public Logic.Pos searchPos(Logic logic, Logic.CellType obj) {
+        Logic.Pos pos = new Logic.Pos(0, 0);
+
+        for (int y = 0; y < logic.getFieldHeight(); y++) {
+            for (int x = 0; x < logic.getFieldWidth(); x++) {
+                if (logic.getCell(x, y).type == obj) {
+                    pos.x = x;
+                    pos.y = y;
+                }
+            }
+        }
+        return pos;
+    }
+
+    private String encodeShadow(final Logic logic) {
+        StringBuilder acc = new StringBuilder();
+
+        for (int y = 0; y < logic.getFieldHeight(); y++) {
+            for (int x = 0; x < logic.getFieldWidth(); x++) {
+                if (logic.getCell(x, y).hasShadow) {
+                    acc.append("x");
+                } else if (logic.getCell(x, y).type == Logic.CellType.TREASURE) {
+                    acc.append("$");
+                } else {
+                    acc.append("o");
+                }
+            }
+            acc.append("\n");
+        }
+        return acc.toString();
+    }
+
+    private Logic.CellType[][] genEmptyField(final int width, final int height) {
+        final Logic.CellType[][] field = new Logic.CellType[height][width];
+
+        for (int y = 0; y < height; y++) {
+            for (int x = 0; x < width; x++) {
+                field[y][x] = Logic.CellType.FLOOR;
+            }
+        }
+        return field;
+    }
+
+    private boolean[][] shadowCoverage(final Logic logic, final int width, final int height) {
+        boolean[][] shadowField = new boolean[height][width];
+        Logic.Cell[][] field = new Logic.Cell[height][width];
+        for (int y = 0; y < height; y++) {
+            for (int x = 0; x < width; x++) {
+                shadowField[y][x] = field[y][x].hasShadow;
+            }
+        }
+        return shadowField;
```

Comment:
Здесь не используется поле из `logic`, а создаётся локальная переменная, не связанная никак с классом `Logic`. 

```suggestion
        boolean[][] shadowField = new boolean[height][width];
        
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                shadowField[y][x] = logic.getCell(x, y).hasShadow;
            }
        }
        return shadowField;
```